stablehlo/compatibility/transforms/CompatibilityPasses.cpp

```diff
@@ -0,0 +1,621 @@
+    // supported version range.
+    // Example:
+    //   CustomCallv1 0.0.0 -> 0.0.x
+    //   CustomCallv2 0.1.0 -> 0.4.x
+    //   CustomCallv3 0.5.0 -> Curr
+    // Target Curr (0.5.0):
+    //   v3 legal    { Curr  in [0.5, Curr] }
+    //   v2 illegal  { Curr !in [0.1, 0.4] }
+    //   v1 illegal  { Curr !in [0.0, 0.0] }
+    // Target 0.4.0:
+    //   v3 illegal { 0.4 !in [0.5, Curr] }
+    //   v2 legal   { 0.4  in [0.1, 0.4] }
+    //   v1 illegal { 0.4 !in [0.0, 0.0] }
+    // Target 0.0.0:
+    //   v3 illegal { 0.0 !in [0.5, Curr] }
+    //   v2 illegal { 0.1 !in [0.1, 0.4] }
+    //   v1 legal   { 0.0  in [0.0, 0.1] }
+    target.addDynamicallyLegalDialect<VersionedhloDialect>([&targetVersionNumber](
+                                                               Operation* op) {
+      if (auto interface = dyn_cast<VersionInterface>(op)) {
+        return (interface.getMinVersion() <= targetVersionNumber &&
+                targetVersionNumber <= interface.getMaxVersion());
+      }
+      return false;
+    });
+
+    versionedhlo::VersionedTypeConverterBase converter;
+    RewritePatternSet patterns(&getContext());
+    versionedhlo::populateVersionedhloToVersionPatterns(&patterns, &converter,
+                                                        &getContext());
+
+    // Conversion from VersionedHLO to StableHLO should never fail.
+    if (failed(applyPartialConversion(getOperation(), target,
+                                      std::move(patterns)))) {
+      return signalPassFailure();
+    }
+  }
+};
+
+////////////////////////////////////////////
+/// Upgrade and Downgrade Infrastructure ///
+////////////////////////////////////////////
+
+LogicalResult emitToVersionError(Operation* op, llvm::StringRef message) {
+  return op->emitError("failed to downgrade ")
+         << op->getName() << ", " << message;
+}
+
+template <typename SourceOp, typename TargetOp>
+struct VersionConversionPattern : OpConversionPattern<SourceOp> {
+  using OpConversionPattern<SourceOp>::OpConversionPattern;
+
+  // This method allows subclasses to add or remove attributes if needed.
+  // Can also fail if an op uses a feature that cannot be represented
+  // in previous versions of the opset.
+  virtual LogicalResult prepareOpForConversion(SourceOp op) const = 0;
+
+  virtual LogicalResult matchAndRewrite(
+      SourceOp op, typename SourceOp::Adaptor adaptor,
+      ConversionPatternRewriter& rewriter) const override {
+    if (failed(prepareOpForConversion(op))) {
+      return failure();
+    }
+    auto newOp = rewriter.replaceOpWithNewOp<TargetOp>(
+        op, op->getResultTypes(), op.getOperands(), op->getAttrs());
+    for (auto [hloRegion, stablehloRegion] :
+         llvm::zip(op->getRegions(), newOp->getRegions())) {
+      rewriter.inlineRegionBefore(hloRegion, stablehloRegion,
+                                  stablehloRegion.end());
+    }
+    return success();
+  }
+};
```

Comment:
I structured it this way because the majority of upgrades (historically) will be for default values attrs which can call the default builder methods. Structuring this way allows for very simple code, and taking it a step further, I had imagined making success the default implementation for prepareOpForConversion, meaning we can write most of our upgrades as:

Currently, I think we can accommodate all upgrades in the PR queue using:

```
struct CustomCallOpV1ToV2
    : public VersionConversionPattern<CustomCallOpV1, CustomCallOpV2> {
  using VersionConversionPattern::VersionConversionPattern;
};

struct CustomCallOpV2ToV3
    : public VersionConversionPattern<CustomCallOpV2, CustomCallOpV3> {
  using VersionConversionPattern::VersionConversionPattern;
};

struct AllGatherOpV1ToV2
    : public VersionConversionPattern<AllGatherOpV1, AllGatherOpV2> {
  using VersionConversionPattern::VersionConversionPattern;
};

struct CollectivePermuteOpV1ToV2
    : public VersionConversionPattern<CollectivePermuteOpV1, CollectivePermuteOpV2> {
  using VersionConversionPattern::VersionConversionPattern;
};
```

Moves/Renames can be modeled with:

```
// Upgrade
struct OriginalNameToNewName
    : public VersionConversionPattern<OriginalName, NewName> {
  using VersionConversionPattern::VersionConversionPattern;
};

// Downgrade
struct NewNameToOriginalName
    : public VersionConversionPattern<NewName, OriginalName> {
  using VersionConversionPattern::VersionConversionPattern;
};
```

Attribute renames and new attrs that don't have a default value can be fixed up using prepareOpForConversion. I'd almost go as far as marking `matchAndRewrite` as `final`.

I suppose we could also go the free function route and leverage a convenience macro as well to build these super-common cases. That would be similar to how IREE does it IIRC.